{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 4:  Growth Rate and Allostery Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook imports a series of mutant counts (already corrected for sequencing error by a Hamming filter), and computes relative growth rates and allosteric effects. The data is then saved as .txt files and pickle databases. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.mlab as mlab\n",
    "import matplotlib\n",
    "import pandas as pd\n",
    "import pickle \n",
    "from scipy import stats \n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.SeqUtils import seq3\n",
    "from Bio.SeqUtils import seq1\n",
    "from matplotlib import rcParams\n",
    "import csv\n",
    "import itertools\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from pathlib import Path #introduced in Python 3.4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Initialize experiment parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the wt sequence of DHFR\n",
    "dhfr_wt = 'MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLNKPVIMGRHTWESIGRPLPGRKNIILSSQPGTDDRVTWVKSVDEAIAACGDVPEIMVIGGGRVYEQFLPKAQKLYLTHIDAEVEGDTHFPDYEPDDWESVFSEFHDADAQNSHSYCFEILERR'\n",
    "dhfr_wt_ix = np.arange(len(dhfr_wt[:]))\n",
    "\n",
    "#all amino acids\n",
    "aas = 'ACDEFGHIKLMNPQRSTVWY*'\n",
    "aas_ix = np.arange(len(aas[:]))\n",
    "\n",
    "#the experiment\n",
    "SL_comp = [list(range(40)),list(range(40,80)),list(range(80,120)),list(range(120,159))] #cutoffs between sublibraries. e.g. SL1=range(0,40)\n",
    "SL_name = ['SL1','SL2','SL3','SL4'] #names of sublibraries \n",
    "vials = ['V1','V2','V3','V4','V5','V6']  #the six vial positions in the turbidostat\n",
    "lit_vials = ['V1','V2','V3'] #these vials were lit during the experiments\n",
    "dark_vials = ['V4','V5','V6'] #these vials were kept in the dark\n",
    "vial_timepoints = [1,2,3,4,5,6] #number of timepoints\n",
    "act_time = [0, 4, 8, 12, 16, 20, 24]#hours from start of experiment that the timepoints were taken\n",
    "\n",
    "#dimensions of stuff\n",
    "nTP = len(act_time)\n",
    "nSL = len(SL_comp)\n",
    "n_pos = len(dhfr_wt)\n",
    "n_aa = len(aas)\n",
    "n_vials = len(vials)\n",
    "n_rpts = len(lit_vials)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Data Import, Growth Rate fitting, and computing allosteric effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Importing all of the data from text files.** Each vial is represented as a three dimensional matrix of 159(pos)\\*21(aa)\\*6(timepoints). Vials 1-3 are repeats in the light, Vials 4-6 are repeats in the dark. All six vials share a single t=0 time point (were inoculated from the same starting culture). The set of six 3D matrices then is compiled into a single dictionary, containing all the raw counts. <br>\n",
    "We also create a matrix of WT counts, which is 4(sublibraries) * 6(timepoints); this gets used in future normalization steps. Data were previously processed by the script: [insert script name(s) here] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_suffix='hard_adjusted_residue.txt' \n",
    "vial_count_matrices = {}\n",
    "WT_counts = np.zeros((nSL,nTP)) #matrix for storing the WT values for each sample (a vial*timepoint combo)\n",
    "path_to_input = Path(\"../Step_3_hamming_filtering/output/\") #OS agnostic pathing\n",
    "\n",
    "#load t=0 (shared for all vials)\n",
    "t0_counts = np.zeros((n_pos,n_aa))#make the array to store the counts in the T0 timepoint\n",
    "f_name_t0 = 'T0'+file_suffix \n",
    "filepath_t0 = path_to_input / f_name_t0 #set up absolute path so we can go to previous output\n",
    "flines = open(filepath_t0,'r').readlines()\n",
    "for line in flines:\n",
    "    sp_line = line.split()\n",
    "    if sp_line[1] == 'WT':\n",
    "        SLix = SL_name.index(sp_line[0])\n",
    "        WT_counts[SLix,0] = float(sp_line[2])\n",
    "    elif sp_line[1] == 'SNEAKY_WT': #silent mutations that encode wildtype, they are thrown out here \n",
    "        continue\n",
    "    else:\n",
    "        pos_ix = int(sp_line[1][1:-1])-1\n",
    "        res_ix = aas.index(sp_line[1][-1])\n",
    "        counts = 0 if float(sp_line[2]) < 1 else float(sp_line[2])\n",
    "        t0_counts[pos_ix,res_ix] = counts     \n",
    "\n",
    "#load all other timepoints\n",
    "for v in vials:\n",
    "    vial_counts = np.zeros((n_pos,n_aa,nTP))\n",
    "    vial_counts[:,:,0] = t0_counts\n",
    "    for tp_ix in vial_timepoints:\n",
    "        f_name = 'T'+str(tp_ix)+v+file_suffix\n",
    "        filepath_timepoint = path_to_input / f_name\n",
    "        flines = open(filepath_timepoint,'r').readlines()\n",
    "        for line in flines:\n",
    "            sp_line = line.split()\n",
    "            if sp_line[1] == 'WT':\n",
    "                SLix = SL_name.index(sp_line[0])\n",
    "                WT_counts[SLix,tp_ix] = float(sp_line[2])\n",
    "            elif sp_line[1] != 'SNEAKY_WT': #silent mutations that encode wildtype, they are thrown out here \n",
    "                pos_ix = int(sp_line[1][1:-1])-1\n",
    "                res_ix = aas.index(sp_line[1][-1])\n",
    "                counts = 0 if float(sp_line[2]) < 1 else float(sp_line[2])\n",
    "                vial_counts[pos_ix,res_ix,tp_ix] = counts\n",
    "    vial_count_matrices[v] = vial_counts \n",
    "    \n",
    "#go back through and fill in WT values in all matrices\n",
    "#this makes sure that all zero values in the matrix truly indicate missing data/absent reads\n",
    "for v in vials:\n",
    "    for p in range(n_pos):\n",
    "        SLix = [p in sl for sl in SL_comp].index(True)\n",
    "        wt_aa_ix = aas.index(dhfr_wt[p])\n",
    "        for tp_ix in range(nTP):\n",
    "            vial_count_matrices[v][p,wt_aa_ix,tp_ix] = WT_counts[SLix,tp_ix]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Normalize the counts to produce relative frequencies**. <br>Produce another set of numpy matrices in which all counts are normalized to a relative frequency $f_a$ as follows: <br>\n",
    "$ f_a = log(\\frac{f_a}{f_{WT}})_{t} - log(\\frac{f_a}{f_{WT}})_{t=0}$\n",
    "\n",
    "<br>\n",
    "Note that this code will produce some warnings, because all mutations with zero counts will produce a $-\\infty$ upon division."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\s178464\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py:7: RuntimeWarning: divide by zero encountered in log\n",
      "  import sys\n",
      "C:\\Users\\s178464\\AppData\\Local\\Continuum\\anaconda3\\lib\\site-packages\\ipykernel_launcher.py:7: RuntimeWarning: invalid value encountered in subtract\n",
      "  import sys\n"
     ]
    }
   ],
   "source": [
    "vial_count_norm_mat = {}\n",
    "for v in vials:\n",
    "    vial_counts_norm = np.zeros(vial_count_matrices[v].shape)\n",
    "    for p in range(n_pos):\n",
    "        SLix = [p in sl for sl in SL_comp].index(True)\n",
    "        for tp_ix in range(nTP):\n",
    "            vial_counts_norm[p,:,tp_ix] = np.log(vial_count_matrices[v][p,:,tp_ix]/WT_counts[SLix,tp_ix]) - \\\n",
    "                                            np.log(vial_count_matrices[v][p,:,0]/WT_counts[SLix,0])\n",
    "    vial_count_norm_mat[v] = vial_counts_norm\n",
    "    #2 warnings are expected: divide by zero, and invalid value. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Compute growth rates.**  \n",
    " We now compute relative growth rates from the matrix of relative frequencies. These growth rates are stored as two three dimensional matrices, of size(159(pos)\\*21(aa)\\*3(repeats)). There is one 3D matrix containing all of the \"lit\" replicates; and one containing all the \"dark\" replicates. <br>\n",
    "\n",
    "Growth rates are computed by linear regression of $f_a$ vs time. In order to fit a growth rate, there must be > T0_gate counts present at t=0. Growth rate is only fit for data points up to the first point where counts go to zero ($f_a = -\\infty $), with a minimum of four points required for fitting.\n",
    "\n",
    "Alongside these two matrices, we create a second pair of \"masking\" matrices, which encode information about missing data and WT positions. More specifically, 0==position ok, 1==position missing at t=0, 2==position present at t=0, but not at sufficient quantities to allow fitting later on, 3==WT position\n",
    "\n",
    "Finally, we condense the data for all replicates into a set of average growth rates (and associated standard deviations) in the lit and dark. Here, we only average over points where fitting was possible (\"0\" in the masking matrix). If computing an average growth rate is not possible (none of the repeats had sufficient data to permit fitting) we mark that position as having a relative growth rate of -1000 (which should never actually occur in these data). In the case that only one repeat permits fitting, we take this value as the average (here it is less clear what to do about the SD... for now I leave it zero). Also record a matrix of the number of points used for averaging for both lit and dark growth rates (to use in assessing variance later)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First, some parameter and function definitions.\n",
    "T0gate = 50\n",
    "\n",
    "#function for computing growth rates. \n",
    "#input parameters:\n",
    "#mat_counts= a npos*naa*ntimepoints matrix of raw counts, \n",
    "#mat_normcounts = associated npos*naa*ntimepoints matrix of normalized counts.\n",
    "#returns:\n",
    "#grates = a npos*naa matrix of fit growth rates, \n",
    "#mask = an associated npos*naa \"mask\" matrix\n",
    "def fit_gr_all(mat_counts,mat_normcounts):\n",
    "    grates = np.zeros([n_pos,n_aa])\n",
    "    mask = np.zeros([n_pos,n_aa])\n",
    "    for p in range(n_pos):\n",
    "        for a,res in enumerate(aas):\n",
    "            #was if mat_counts[p,a,0] <= T0gate:\n",
    "            if mat_counts[p,a,0] < T0gate:\n",
    "                if mat_counts[p,a,0] < T0gate and mat_counts[p,a,0] > 0:\n",
    "                    mask[p,a] = 2\n",
    "                else:\n",
    "                    mask[p,a] = 1\n",
    "            else:\n",
    "                #checking to see which timepoints have more than zero counts\n",
    "                if (False in list(mat_counts[p,a,:] > 1)):\n",
    "                    stop_fit_ix = list(mat_counts[p,a,:] > 1).index(False) \n",
    "                else:\n",
    "                    stop_fit_ix = len(mat_counts[p,a,:])\n",
    "                    \n",
    "                if stop_fit_ix < 3:\n",
    "                    #not enough data to fit\n",
    "                    mask[p,a] = 2\n",
    "                else:\n",
    "                    m,b,r,pval,se = stats.linregress(act_time[0:stop_fit_ix],mat_normcounts[p,a,0:stop_fit_ix])\n",
    "                    grates[p,a] = m\n",
    "            if res == dhfr_wt[p]:\n",
    "                mask[p,a] = 3          \n",
    "    return grates,mask\n",
    "\n",
    "#function for averaging growth rates across repeats.\n",
    "#input parameters:\n",
    "#gr_mat = npos*naa*nrpts matrix of growth rates\n",
    "#mask_mat = associated npos*naa*nrpts masking matrix (that we use to determine which points to average)\n",
    "#the function returns:\n",
    "#avg_mat = npos*naa matrix of average growth rates\n",
    "#std_mat = npos*naa matrix of std deviations in growth rates\n",
    "#n_samp_mat = npos*naa matrix of number of measurements used to compute each average\n",
    "def avg_gr_replicates(gr_mat,mask_mat):\n",
    "    avg_mat,std_mat,n_samp_mat = np.zeros([n_pos,n_aa]),np.zeros([n_pos,n_aa]),np.zeros([n_pos,n_aa])\n",
    "    for p in range(n_pos):\n",
    "        for a in range(n_aa):\n",
    "            to_avg = [gr_mat[p,a,k] for k,mval in enumerate(mask_mat[p,a]) if mval==0]\n",
    "            n_samp_mat[p,a]=len(to_avg)\n",
    "            if len(to_avg) == 0:\n",
    "                to_avg_quick_death = [gr_mat[p,a,k] for k,mval in enumerate(mask_mat[p,a]) if mval==2]\n",
    "                \n",
    "                #changing this from <= 1 to <= 0, aka = 0. \n",
    "                if len(to_avg_quick_death) <= 0:\n",
    "                    avg_mat[p,a] = -1000\n",
    "                else: avg_mat[p,a] = -999   \n",
    "            elif len(to_avg) == 1:\n",
    "                #here changed to -999 from to_avg[0]\n",
    "                avg_mat[p,a] = -999\n",
    "            else:\n",
    "                avg_mat[p,a] = np.mean(to_avg)\n",
    "                std_mat[p,a] = np.std(to_avg)\n",
    "    return avg_mat,std_mat,n_samp_mat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The actual calculations.\n",
    "#compute the growth rates and masks for all lit and dark vials\n",
    "all_gr_lit, all_gr_dark = np.zeros([n_pos,n_aa,n_rpts]), np.zeros([n_pos,n_aa,n_rpts])\n",
    "all_gr_lit_mask, all_gr_dark_mask = np.zeros([n_pos,n_aa,n_rpts]), np.zeros([n_pos,n_aa,n_rpts])\n",
    "for i,v in enumerate(lit_vials):\n",
    "    gr_tmp,mask_tmp = fit_gr_all(vial_count_matrices[v],vial_count_norm_mat[v])\n",
    "    all_gr_lit[:,:,i] = gr_tmp\n",
    "    all_gr_lit_mask[:,:,i] = mask_tmp\n",
    "    \n",
    "for i,v in enumerate(dark_vials):\n",
    "    gr_tmp,mask_tmp = fit_gr_all(vial_count_matrices[v],vial_count_norm_mat[v])\n",
    "    all_gr_dark[:,:,i] = gr_tmp\n",
    "    all_gr_dark_mask[:,:,i] = mask_tmp\n",
    "    \n",
    "#compute average and std error across repeats for each condition\n",
    "avg_gr_lit, sd_gr_lit, n_pts_gr_lit = avg_gr_replicates(all_gr_lit,all_gr_lit_mask)\n",
    "avg_gr_dark, sd_gr_dark, n_pts_gr_lit = avg_gr_replicates(all_gr_dark,all_gr_dark_mask)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Compute allostery** Given the above set of growth rates, we can now compute the allosteric effect as the difference between the average growth rate in the light, and the average growth rate in the dark. <br>\n",
    "\n",
    "To indicate mutations where allostery cannot be reliably computed (because one or both average growth rates could not be computed), we mark the allosteric effect as -1000 (which is otherwise not possible in this dataset)\n",
    "\n",
    "As a measure of the significance of this difference, we also compute a p-value by t-test, which tells us the probability that the means come from the same distribution (using Welch's t-test, which does not assume an equal variance in both samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "allostery = avg_gr_lit - avg_gr_dark\n",
    "allostery_pvals = np.ones([n_pos,n_aa])\n",
    "allostery_sd = np.zeros([n_pos,n_aa])\n",
    "\n",
    "#In this loop:\n",
    "#1) set allostery = -1000 for mutations lacking growth rate data\n",
    "#2) calculate pvals (by Welch's t-test)\n",
    "#3) calculate propagated error (given the underlying growth rate data)\n",
    "for p in range(n_pos):\n",
    "    for a in range(n_aa):\n",
    "\n",
    "        if (avg_gr_lit[p,a] == -1000) or (avg_gr_dark[p,a] == -1000):\n",
    "            allostery[p,a] = -1000 \n",
    "            #These values are from DL121D27N, which is considered catalytically dead.\n",
    "        if ((avg_gr_lit[p,a] <= -0.1301) or (avg_gr_dark[p,a] <= -0.1329)) and not\\\n",
    "        ((avg_gr_lit[p,a] == -1000) or (avg_gr_dark[p,a] == -1000)):\n",
    "            allostery[p,a] = -999\n",
    "            \n",
    "        lit_vals = [all_gr_lit[p,a,k] for k,mval in enumerate(all_gr_lit_mask[p,a]) if mval==0]\n",
    "        dark_vals = [all_gr_dark[p,a,k] for k,mval in enumerate(all_gr_dark_mask[p,a]) if mval==0]\n",
    "        if len(lit_vals)>1 and len(dark_vals) > 1:\n",
    "            t,pval = stats.ttest_ind(lit_vals,dark_vals,equal_var=False)\n",
    "            allostery_pvals[p,a] = pval\n",
    "        \n",
    "        allostery_sd[p,a] = (sd_gr_lit[p,a]**2 + sd_gr_dark[p,a]**2)**0.5\n",
    "        \n",
    "        \n",
    "#we do the same but without filtering out the catalytically dead mutations        \n",
    "\n",
    "allostery_all = avg_gr_lit - avg_gr_dark\n",
    "allostery_pvals_all = np.ones([n_pos,n_aa])\n",
    "allostery_sd_all = np.zeros([n_pos,n_aa])\n",
    "\n",
    "for p in range(n_pos):\n",
    "    for a in range(n_aa):\n",
    "\n",
    "        if (avg_gr_lit[p,a] == -1000) or (avg_gr_dark[p,a] == -1000):\n",
    "            allostery_all[p,a] = -1000 \n",
    "        if ((avg_gr_lit[p,a] <= -999.0) or (avg_gr_dark[p,a] <= -999.0)) and not\\\n",
    "        ((avg_gr_lit[p,a] == -1000) or (avg_gr_dark[p,a] == -1000)):\n",
    "            allostery_all[p,a] = -999\n",
    "            \n",
    "        lit_vals_all = [all_gr_lit[p,a,k] for k,mval in enumerate(all_gr_lit_mask[p,a]) if mval==0]\n",
    "        dark_vals_all = [all_gr_dark[p,a,k] for k,mval in enumerate(all_gr_dark_mask[p,a]) if mval==0]\n",
    "        if len(lit_vals)>1 and len(dark_vals) > 1:\n",
    "            t,pval_all = stats.ttest_ind(lit_vals_all,dark_vals_all,equal_var=False)\n",
    "            allostery_pvals_all[p,a] = pval_all\n",
    "        \n",
    "        allostery_sd_all[p,a] = (sd_gr_lit[p,a]**2 + sd_gr_dark[p,a]**2)**0.5        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **This concludes the calculations portion of the code**. \n",
    "Writing all results to disk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def picklefiler(variable_name, filename):\n",
    "    outfile = open(filename,'wb')\n",
    "    pickle.dump(variable_name,outfile)\n",
    "    outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ix,v in enumerate(lit_vials): #growth rate for each lit vial separately \n",
    "    np.savetxt('output/'+v+'.txt', all_gr_lit[:,:,ix] ,delimiter=',') #V1, V2, V3\n",
    "    \n",
    "picklefiler(all_gr_lit,'output/all_gr_lit.pkl') #growth rates for individual replicates in light\n",
    "\n",
    "for ix,v in enumerate(dark_vials): #growth rate for each dark vial separately \n",
    "    np.savetxt('output/'+v+'.txt', all_gr_dark[:,:,ix] ,delimiter=',') #V4, V5, V6\n",
    "    \n",
    "picklefiler(all_gr_dark,'output/all_gr_dark.pkl')#growth rates for individual replicates in dark  \n",
    "\n",
    "picklefiler(t0_counts,'output/t0_counts.pkl')#counts at t=0 \n",
    "\n",
    "np.savetxt('output/lit_gr_avg.txt', avg_gr_lit, delimiter=',')  \n",
    "picklefiler(avg_gr_lit,'output/avg_gr_lit.pkl') #average growth rate in the light for all mutations\n",
    "\n",
    "np.savetxt('output/lit_gr_std.txt', sd_gr_lit, delimiter=',') \n",
    "picklefiler(sd_gr_lit,'output/sd_gr_lit.pkl')#standard deviation of the growth rate in the light\n",
    "\n",
    "np.savetxt('output/dark_gr_avg.txt', avg_gr_dark, delimiter=',') \n",
    "picklefiler(avg_gr_dark,'output/avg_gr_dark.pkl')#average growth rate in the dark for all mutations\n",
    "\n",
    "np.savetxt('output/dark_gr_std.txt', sd_gr_dark, delimiter=',') \n",
    "picklefiler(sd_gr_dark,'output/sd_gr_dark.pkl')#standard deviation of the growth rate in the dark\n",
    "\n",
    "np.savetxt('output/allostery.txt', allostery, delimiter=',') \n",
    "picklefiler(allostery,'output/allostery.pkl')#allostery (lit-dark growth rate)\n",
    "picklefiler(allostery_all,'output/allostery_all.pkl')#allostery not filtering catalytically dead (lit-dark growth rate)\n",
    "\n",
    "np.savetxt('output/allostery_pvals.txt', allostery_pvals, delimiter=',') \n",
    "picklefiler(allostery_pvals,'output/allostery_pvals.pkl')#allostery p-value\n",
    "picklefiler(allostery_pvals_all,'output/allostery_pvals_all.pkl')#allostery p-value not filtering catalytically dead\n",
    "\n",
    "np.savetxt('output/allostery_sd.txt', allostery_sd, delimiter=',')\n",
    "picklefiler(allostery_sd,'output/allostery_sd.pkl')#allostery std deviation  \n",
    "picklefiler(allostery_sd_all,'output/allostery_sd_all.pkl')#allostery std deviation not filtering catalytically dead\n",
    "\n",
    "#dictionary of relative frequency of each mutant at each timepoint and vial\n",
    "picklefiler(vial_count_norm_mat,'output/vial_count_norm_mat.pkl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
